mce-label-injector/webhook/webhook.go

```diff
@@ -0,0 +1,128 @@
+package hook
+
+import (
+	"context"
+	"encoding/json"
+	"github.com/go-logr/logr"
+	"gomodules.xyz/jsonpatch/v2"
+	admissionv1 "k8s.io/api/admission/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/kube-openapi/pkg/validation/errors"
+	"net/http"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+var log = logf.Log.WithName("example-controller")
+
+// +kubebuilder:webhook:path=/mutate,mutating=true,failurePolicy=fail,groups="cluster.x-k8s.io",verbs=create;update,versions=v1,name=mce-label-injector.x-k8s.io
+
+// MceLabelInjector label MCE objects for groups="cluster.x-k8s.io"
+type MceLabelInjector struct {
+	Client         client.Client
+	MceLabelConfig *Config
+	ClientSet      *kubernetes.Clientset
+	Log            logr.Logger
+}
+
+type Config struct {
+	NamespaceOpenshiftClusterApi string `default:"openshift-cluster-api"                        yaml:"namespace_openshift_cluster_api"`
+	HyperShiftLabelName          string `default:"hypershift.openshift.io/hosted-control-plane" yaml:"hyper_shift_label_name"`
+	LabelMultiClusterEngine      string `default:"multicluster-engine"                          yaml:"label_multi_cluster_engine"`
+}
+
+func (li *MceLabelInjector) countLabel(ctx context.Context, namespaceName string, oldValue string) (string, bool, error) {
+	/*
+		* The auto-labeling mutating webhook inspects the NS
+		  * If the namespace is openshift-cluster-api, don't label for MCE
+		  * Else, fetch the NS and inspect labels
+		    * If it has the hypershift label, don't label for MCE
+		  * If the NS is for (HyperShift or openshift-cluster-api) AND already has the MCE label, reject the admission, invalid configuration
+		  * If we've not returned by now, add the MCE label
+		* Relevant controller acts on CR
+		  * MCE using a watchfilter on their label
+	*/
+
+	newLabel := ""
+
+	// If the namespace is openshift-cluster-api, don't label for MCE
+	if namespaceName == li.MceLabelConfig.NamespaceOpenshiftClusterApi {
+		return "", false, nil
+	} else {
+		// Else, fetch the NS and inspect labels
+		namespace, err := li.ClientSet.CoreV1().Namespaces().Get(ctx, namespaceName, metav1.GetOptions{})
+		if err != nil {
+			return "", false, errors.New(0, "Cannot get namespace='%s' : %s", namespaceName, err.Error())
+		}
+		hyperShiftValue, isHyperShift := namespace.Labels[li.MceLabelConfig.HyperShiftLabelName]
+		if isHyperShift && hyperShiftValue == "true" {
+			return "", false, nil
+		}
+	}
+
+	// If we've not returned by now, add the MCE label
+	newLabel = li.MceLabelConfig.LabelMultiClusterEngine
+
+	// If the NS is for (HyperShift or openshift-cluster-api) AND already has the MCE label, reject the admission, invalid configuration
```

Comment:
Not sure if this logic is correct , It already return if the NS is openshift-cluster-api [here](https://github.com/stolostron/cluster-api-installer/pull/37/files#diff-65888fb6ad080d04f36f5d7a7949f10ba23eb1fba42cea42fa29918655f34175R52)